# Importing libraries

In [2]:
import pandas as pd
import numpy as np
import os

# Import files

In [3]:
# Creating path
path = r"C:\Users\azikb\Downloads\Data Immersion. Assiginment 4. Python. Instacart Basket Analysis"

In [4]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# Consistancy check

In [6]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

# Create new Data Frame

In [7]:
df_test = pd.DataFrame()

In [8]:
df_test['mix'] = ['a', 'b', 1, True]

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

# Let's check Data Frame 'test' for mixed data

In [10]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\azikb\AppData\Local\Temp\ipykernel_14280\1786467950.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [11]:
# Let'us change the type of dataframe 'test'
df_test['mix'] = df_test['mix'].astype('str')

# Missing Values

In [12]:
# Finding out the missing falues
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

#### Creating a subset (pointing out) missing data from 'product_name' column

In [13]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# Addressing missing Values


#### Using 'mean' value

In [15]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

#### As i understand, we can not use 'mean' command here, because the 'product_name' column has a non numeric data type.

#### Using 'median' value

In [16]:
df_prods['prices'].median()

7.1

#### I can not use median at 'product name', because it's not a numeric column. As an example i took 'prices' column

# Let's compare the Data Frames 'df_prods' and 'df_prods_clean'

In [17]:
df_prods.shape

(49693, 5)

#### Now, let's create subset 'non-missing' values

In [18]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [19]:
df_prods_clean.shape

(49677, 5)

#### As we can see, the difference is 16 (49693-49677=16) 

# Another way to drop off columns

In [20]:
df_prods.dropna(subset = ['product_name'], inplace = True)

In [21]:
df_prods.shape

(49677, 5)

#### As we can see, the function drop off 16 missing rows from dataframe. Now total number of rows is 49677, the same like in df_prods_clean

# Duplicates

In [22]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [23]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

# Addressing Duplicates

In [24]:
# Before we addressing duplicates, let's see what number of rows in df_prods_clean dataframe
df_prods_clean.shape

(49677, 5)

In [25]:
# Now let's make a subset 'df_prods_clean_no_dups' and drop of the duplicate rows 
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [26]:
df_prods_clean_no_dups.shape

(49672, 5)

#### As we can see, we drop of 5 duplicated rows (49677-5=49672)

# Now, let's EXPORT our new, clean file to folder 'Prepared Data'

In [27]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'product_checked.csv'))

# Task. Directions

#### 2 Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further

In [28]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

#### As we can observe from table above is a basic statistics. We are interested only in 'order_day_of_week', 'order_hour_of_day', 'days_since_prior_order'. The 'max' for odrer _hour_of_day is '2.3', which is strange. The 'max' for column 'days_since_prior_order' looks also strange.In general, the 'ords' file looks ok

#### We have missing data in 'days_since_prior_order'

#### max for the 'days_since_prior_order' looks strange = 3.

# 3 Check for mixed-type data in your df_ords dataframe.

In [29]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

C:\Users\azikb\AppData\Local\Temp\ipykernel_14280\1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\azikb\AppData\Local\Temp\ipykernel_14280\1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\azikb\AppData\Local\Temp\ipykernel_14280\1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\azikb\AppData\Local\Temp\ipykernel_14280\1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\azikb\AppData\Local

#### Non of the columns have mixed data

# 4 If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

#### I do not see any mixed type data

# 5 Run a check for missing values in your df_ords dataframe. In a markdown cell, report your findings and propose an explanation for any missing values you findd.

In [30]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### As we see, missing data is in column 'days_since_prior_order'. We do not have any info about 206209 rows from these columns.This info was not provided. Resons could be: 1.the customer delete account, 2. the customer didn't make any purchase since yet, 3. customer doesn't satisfied with service and doesn't want to make any purchase here...  

# 6 Address the missing values using an appropriate method. In a markdown cell, explain why you used your method of choice..

#### I wouldn't erase the missing data or substitute them with average value or so what...Before doing something with missing data we need to contact to CUSTOMER

In [31]:
# Instead of erasing missing data, we can create a subset with non-missing values
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [32]:
df_nan

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
1                 1   2398795        1             2                  3   
2                 2    473747        1             3                  3   
3                 3   2254736        1             4                  4   
4                 4    431534        1             5                  4   
5                 5   3367565        1             6                  2   
...             ...       ...      ...           ...                ...   
3421078     3421078   2266710   206209            10                  5   
3421079     3421079   1854736   206209            11                  4   
3421080     3421080    626363   206209            12                  1   
3421081     3421081   2977660   206209            13                  1   
3421082     3421082    272231   206209            14                  6   

         order_hour_of_day  days_since_prior_order  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
5                        7                    19.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3214874 rows x 7 columns]

# 7 Run a check for duplicate values in your df_ords data. In a markdown cell, report your findings and propose an explanation for any duplicate values you find..

In [33]:
df_dups = df_ords[df_ords.duplicated()]

In [34]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

#### As we can see we have no duplicates so far...

# 8 Address the duplicates using an appropriate method..

#### I do not have any duplicates...

# 9 Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [35]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

In [36]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))

In [37]:
df_ords.shape

(3421083, 7)